src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -439,9 +441,20 @@ export function generateSamplePoints(
   // Win 11, NVidia 2070 Super: 16
   // Linux, AMD Radeon Pro WX 3200: 256
   // MacOS, M1 Mac: 256
-  const kSubdivisionsPerTexel = 2;
+  const kSubdivisionsPerTexel = 4;
   const q = [args.textureWidth * kSubdivisionsPerTexel, args.textureHeight * kSubdivisionsPerTexel];
-  return out.map(c => c.map((v, i) => Math.floor(v * q[i]) / q[i]) as vec2);
+  return out.map(
+    c =>
+      c.map((v, i) => {
+        // Quantize to kSubdivisionsPerPixel
+        const v1 = Math.floor(v * q[i]);
+        // If it's nearest and we're on the edge of a texel then move us off the edge
```

Comment:
I actually implemented something similar to deal with 'bumping' edge texels - apparently I reverted it, or have a local branch somewhere I never pushed.

Please correct me if I'm wrong, but I think what you have here will bump the texels between `[0, ε)`, but you have to consider both sides of the edge (`[-ε, ε]`). IIRC my approach was something like:

```ts
// Bump all texels that sit on a texel boundary, so that they're deterministically sampling.
out = out.map(c => c.map((v, i) => {
  const dims = [args.textureWidth, args.textureHeight];
  const t = v * dims[i]; // normalized -> integer texels
  const onEdge = Math.abs(fract(t + 0.5) - 0.5) < epsilon;
  if (onEdge) {
    t += bumpFactor;
  }
  return t / dims[i];
})
```

Reply:
The code as is, works

```
nearest = true;
args = { textureWidth: 8, textureHeight: 8};
kSubdivisionsPerTexel = 4;
q = [args.textureWidth * kSubdivisionsPerTexel, args.textureHeight * kSubdivisionsPerTexel];
bumpFromTexelBoundry = (v, i) => {
      // Quantize to kSubdivisionsPerPixel
      const v1 = Math.floor(v * q[0]);
      // If it's nearest and we're on the edge of a texel then move us off the edge
      // since the edge could choose one texel or another in nearest mode
      const v2 = (nearest && v1 % kSubdivisionsPerTexel === 0) ? v1 + 1 : v1;
      // Convert back to texture coords
      return v2 / q[0];
}
toTexel = v => v * args.textureWidth;
toCoord = v => v / args.textureWidth;
[-2, -1.5, -1, -0.5, 0, 0.5, 1, 1.5, 2].map(toCoord).map(bumpFromTexelBoundry).map(toTexel)
```

> `(9) [-1.75, -1.5, -0.75, -0.5, 0.25, 0.5, 1.25, 1.5, 2.25]`

Unfortunately it didn't help get the tests to pass on android

Reply:
I believe it works because it's quantized first. Basically you're just mapping
`[0, 1, 2, 3, 4, 5, 6, 7, ...]` to
`[1, 1, 2, 3, 5, 5, 6, 7, ...]`